core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableTableScan.java

```diff
@@ -84,15 +90,59 @@ public static EnumerableTableScan create(RelOptCluster cluster,
   }
 
   /** Returns whether EnumerableTableScan can generate code to handle a
-   * particular variant of the Table SPI. */
+   * particular variant of the Table SPI.
+   * @deprecated
+   **/
+  @Deprecated
   public static boolean canHandle(Table table) {
-    // FilterableTable and ProjectableFilterableTable cannot be handled in
-    // enumerable convention because they might reject filters and those filters
-    // would need to be handled dynamically.
+    if (table instanceof TransientTable) {
+      // CALCITE-3673: TransientTable can't be implemented with Enumerable
+      return false;
+    }
+    // See org.apache.calcite.prepare.RelOptTableImpl.getClassExpressionFunction
     return table instanceof QueryableTable
+        || table instanceof FilterableTable
+        || table instanceof ProjectableFilterableTable
         || table instanceof ScannableTable;
   }
 
+  /** Returns whether EnumerableTableScan can generate code to handle a
+   * particular variant of the Table SPI.
+   **/
+  public static boolean canHandle(RelOptTable relOptTable) {
+    Table table = relOptTable.unwrap(Table.class);
+    if (table != null && !canHandle(table)) {
+      return false;
+    }
+    boolean supportArray = CalciteSystemProperty.ENUMERABLE_ENABLE_TABLESCAN_ARRAY.value();
+    boolean supportMap = CalciteSystemProperty.ENUMERABLE_ENABLE_TABLESCAN_MAP.value();
+    boolean supportMultiset = CalciteSystemProperty.ENUMERABLE_ENABLE_TABLESCAN_MULTISET.value();
+    if (supportArray && supportMap && supportMultiset) {
+      return true;
+    }
+    // Struct fields are not supported in EnumerableTableScan
+    for (RelDataTypeField field : relOptTable.getRowType().getFieldList()) {
+      boolean unsupportedType = false;
+      switch (field.getType().getSqlTypeName()) {
+      case ARRAY:
+        unsupportedType = supportArray;
+        break;
+      case MAP:
```

Comment:
Better move the type decision to `SqlTypeUtil`.

Reply:
I guess this is already in master.
Will rebase this PR.